plugins/mongodb/mongodbembed.hpp

```diff
@@ -0,0 +1,695 @@
+        /**
+         * @brief Returns a pointer to the result rows from the MongoDB operation.
+         * 
+         * @return StringArray* Result rows in standard json.
+         */
+        StringArray* result() 
+        {
+            return &result_rows;
+        }
+
+        /**
+         * @brief Returns a copy of the batch size to use in mongocxx::options
+         * 
+         * @return std::int32_t Batch Size
+         */
+        std::int32_t size()
+        {
+            return batchSize;
+        }
+
+    protected:
+        std::string databaseName;                      //! Local copy of database name.
+        std::string collectionName;                    //! Local copy of collection name.
+        std::string embeddedScript;                    //! Local copy of the embedded script.
+        std::string queryCMD;                          //! Local copy of the name of the MongoDB query operation.
+        bsoncxx::builder::basic::document builder{};   //! Local copy of the basic builder used for bind ECL rows to documents.
+        const char* cursor;                            //! Pointer for keeping track of parsing the embedded script.
+        StringArray result_rows;                       //! Local copy of result rows.
+        std::int32_t batchSize;                        //! Batch Size for result rows.
+    };
+
+    /**
+     * @brief Builds ECL Records from MongoDB result rows.
+     * 
+     */
+    class MongoDBRowStream : public RtlCInterface, implements IRowStream
+    {
+    public:
+        MongoDBRowStream(IEngineRowAllocator* _resultAllocator, std::shared_ptr<MongoDBQuery> _query);
+        virtual ~MongoDBRowStream();
+
+        RTLIMPLEMENT_IINTERFACE
+        virtual const void* nextRow();
+        virtual void stop();
+    private:
+        Linked<IEngineRowAllocator> m_resultAllocator;      //!< Pointer to allocator used when building result rows.
+        bool m_shouldRead;                                  //!< If true, we should continue trying to read more messages.
+        __int64 m_currentRow;                               //!< Current result row.
+        std::shared_ptr<MongoDBQuery> m_query;                              //!< Pointer to MongoDBQuery object.
+    };
+
+    /**
+     * @brief Class for keeping a heap allocated MongoDB instance. The reason this is important is because
+     * MongoDB only allows one instance to be created at any one time, so we create an instance on the heap
+     * for every thread to share.
+     * 
+     */
+    class MongoDBConnection 
+    {
+    public:
+        /**
+         * @brief Creates a static reference to a MongoDB instance that is alive
+         * for the entire time MongoDBEmbedFunctionContext is used.
+         * 
+         * @return MongoDBConnection& A reference to a MongoDB instance used for
+         * connecting to a database.
+         */
+        static MongoDBConnection& instance() 
+        {
+            static MongoDBConnection instance;
```

Comment:
This is one of the nuances of MongoDB. It is very explicit in that it only allows a single instance object to be alive at one time. The only place I have been able to find where you can pass in connection options is in the instance constructor. The issue is that there is zero documentation on this and I have only been able to find a very simple example of these objects being used from MongoDB. I also feel as though there should be a way for each connection to have their own options even if they share an instance object I just haven't been able to find any information about it. This is something that will require more research and I will look into it more in depth.

Reply:
My reading of your code was that the configuration was supplied to the pool rather than the mongodb instance. So can you have multiple different pool instances that share the MongoDBConnection?  Beyond the scope of this PR, but please can you create a jira and add the number here so it can be revisited later.

Reply:
That is my thought as well, but when I was starting the plugin I couldn't find any information about how that works with the pool objects. I have created a jira issue here: https://track.hpccsystems.com/browse/HPCC-28326